// web/src/ui/Card.tsx
import { forwardRef } from 'react';
import { cn } from './cn';

type CardVariant = 'default' | 'elevated' | 'bordered' | 'glass';

type CardProps = React.HTMLAttributes<HTMLDivElement> & {
  variant?: CardVariant;
  interactive?: boolean;
  padding?: 'none' | 'sm' | 'md' | 'lg';
};

type CardHeaderProps = React.HTMLAttributes<HTMLDivElement> & {
  title?: string;
  subtitle?: string;
  right?: React.ReactNode;
};

type CardBodyProps = React.HTMLAttributes<HTMLDivElement>;

type CardFooterProps = React.HTMLAttributes<HTMLDivElement>;

const variantClasses = {
  default: 'bg-white border border-slate-200 shadow-sm',
  elevated: 'bg-white shadow-lg border border-slate-100',
  bordered: 'bg-white border-2 border-slate-200',
  glass: 'bg-white/80 backdrop-blur-md border border-white/20 shadow-lg',
};

const paddingClasses = {
  none: '',
  sm: 'p-3',
  md: 'p-4',
  lg: 'p-6',
};

export const Card = forwardRef<HTMLDivElement, CardProps>(
  ({ className, variant = 'default', interactive = false, padding = 'md', children, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn(
          'rounded-2xl transition-all duration-200',
          variantClasses[variant],
          paddingClasses[padding],
          interactive && 'cursor-pointer hover:-translate-y-1 hover:shadow-xl active:translate-y-0',
          className
        )}
        {...props}
      >
        {children}
      </div>
    );
  }
);

Card.displayName = 'Card';

export const CardHeader = forwardRef<HTMLDivElement, CardHeaderProps>(
  ({ className, title, subtitle, right, children, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn('flex items-start gap-3 pb-4 border-b border-slate-100', className)}
        {...props}
      >
        {(title || subtitle) && (
          <div className="min-w-0 flex-1">
            {title && <h3 className="text-base font-bold text-slate-900 leading-tight">{title}</h3>}
            {subtitle && <p className="mt-1 text-sm text-slate-600">{subtitle}</p>}
          </div>
        )}
        {children}
        {right && <div className="ml-auto shrink-0">{right}</div>}
      </div>
    );
  }
);

CardHeader.displayName = 'CardHeader';

export const CardBody = forwardRef<HTMLDivElement, CardBodyProps>(
  ({ className, children, ...props }, ref) => {
    return (
      <div ref={ref} className={cn('py-4', className)} {...props}>
        {children}
      </div>
    );
  }
);

CardBody.displayName = 'CardBody';

export const CardFooter = forwardRef<HTMLDivElement, CardFooterProps>(
  ({ className, children, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn('flex items-center gap-2 pt-4 border-t border-slate-100', className)}
        {...props}
      >
        {children}
      </div>
    );
  }
);

CardFooter.displayName = 'CardFooter';
