// web/src/ui/Input.tsx
import { forwardRef } from 'react';
import { cn } from './cn';

type InputSize = 'sm' | 'md' | 'lg';
type InputVariant = 'default' | 'filled' | 'flushed';

type InputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  label?: string;
  error?: string;
  helperText?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  inputSize?: InputSize;
  variant?: InputVariant;
  fullWidth?: boolean;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      className,
      label,
      error,
      helperText,
      leftIcon,
      rightIcon,
      inputSize = 'md',
      variant = 'default',
      fullWidth = true,
      disabled,
      ...props
    },
    ref
  ) => {
    const sizeClasses = {
      sm: 'h-9 px-3 text-sm',
      md: 'h-11 px-4 text-sm',
      lg: 'h-13 px-5 text-base',
    };

    const variantClasses = {
      default: cn(
        'bg-white border border-slate-200',
        'focus:border-blue-500 focus:ring-4 focus:ring-blue-500/10',
        error && 'border-red-500 focus:border-red-500 focus:ring-red-500/10'
      ),
      filled: cn(
        'bg-slate-100 border border-transparent',
        'focus:bg-white focus:border-blue-500 focus:ring-4 focus:ring-blue-500/10',
        error && 'bg-red-50 border-red-500 focus:border-red-500 focus:ring-red-500/10'
      ),
      flushed: cn(
        'bg-transparent border-b-2 border-slate-200 rounded-none px-0',
        'focus:border-blue-500',
        error && 'border-red-500 focus:border-red-500'
      ),
    };

    return (
      <div className={cn('space-y-1.5', fullWidth && 'w-full')}>
        {label && (
          <label className="block text-sm font-semibold text-slate-700">
            {label}
            {props.required && <span className="ml-1 text-red-500">*</span>}
          </label>
        )}

        <div className="relative">
          {leftIcon && (
            <div className="absolute left-3 top-1/2 -translate-y-1/2 text-slate-400">
              {leftIcon}
            </div>
          )}

          <input
            ref={ref}
            disabled={disabled}
            className={cn(
              'w-full rounded-xl outline-none transition-all duration-200',
              'placeholder:text-slate-400',
              'disabled:cursor-not-allowed disabled:opacity-60 disabled:bg-slate-50',
              sizeClasses[inputSize],
              variantClasses[variant],
              leftIcon && 'pl-10',
              rightIcon && 'pr-10',
              className
            )}
            {...props}
          />

          {rightIcon && (
            <div className="absolute right-3 top-1/2 -translate-y-1/2 text-slate-400">
              {rightIcon}
            </div>
          )}
        </div>

        {(error || helperText) && (
          <p className={cn('text-xs', error ? 'text-red-600' : 'text-slate-500')}>
            {error || helperText}
          </p>
        )}
      </div>
    );
  }
);

Input.displayName = 'Input';
